### PR TITLE
`Development`: Increase client test coverage for http interceptors

### PR DESCRIPTION
**Pull request by coaxial1**
#### General

- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).

### Motivation and Context

Test previously untested interceptors.

#### Code Review
- [x] Review 1
- [x] Review 2

### Test Coverage

|  | Branches | Lines | Remarks
|  -- | -- |-- | -- |
| app/core/interceptor                                                              | 65.88 |   76.19 |  Will be higher after Tests for Version Interceptor are added                                                                            
|  -- artemis-version.interceptor.ts                                            |               0 |       0 |  Tests will be added in #4518                                                              
|  -- auth-expired.interceptor.ts                                                 |            80 |      100 |                                                                     
|  -- auth.interceptor.ts                                                                | 83.33 |       100 |                                                                            
|  -- browser-fingerprint.interceptor.service.ts                          |   88.88 |         100 |                                                                           
|  -- errorhandler.interceptor.ts                                                    |     90 |        100 |                                                                       
|  -- interceptor.util.ts                                                                |    83.33 |        100 |                                                                       
|  -- notification.interceptor.ts                                                        |    88.88 |        100 |

### CODE REVIEW

File: src/test/javascript/spec/interceptor/auth-expired.interceptor.spec.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        jest.restoreAllMocks();
+    });
+
+    it('should logout and store the current url if status is 401 and there is no token', () => {
+        const mockHandler = {
+            handle: () => throwError(() => new HttpErrorResponse({ status: 401 })),
+        };
+        const tokenSpy = jest.spyOn(authServerProviderMock, 'getToken').mockReturnValue('token-123');
+
+        authInterceptor.intercept(null as any, mockHandler).subscribe();
```
Comment (julian-christl) [inline src/test/javascript/spec/interceptor/auth-expired.interceptor.spec.ts:48]: Please use an actual value. Same for other occurrences

File: src/test/javascript/spec/interceptor/auth-expired.interceptor.spec.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    afterEach(() => {
+        jest.restoreAllMocks();
+    });
+
+    it('should logout and store the current url if status is 401 and there is no token', () => {
+        const mockHandler = {
+            handle: () => throwError(() => new HttpErrorResponse({ status: 401 })),
+        };
+        const tokenSpy = jest.spyOn(authServerProviderMock, 'getToken').mockReturnValue('token-123');
```
Comment (julian-christl) [inline src/test/javascript/spec/interceptor/auth-expired.interceptor.spec.ts:46]: Make the token a constant as in the other file(s)

File: src/test/javascript/spec/interceptor/auth.interceptor.spec.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            handle: jest.fn(),
+        };
+
+        authInterceptor.intercept(requestMock, mockHandler);
+
+        expect(localStorageSpy).toHaveBeenCalledWith('authenticationToken');
+        expect(sessionStorageSpy).toHaveBeenCalledWith('authenticationToken');
+        expect(cloneSpy).toHaveBeenCalledTimes(0);
+        expect(mockHandler.handle).toHaveBeenCalledTimes(1);
+        expect(mockHandler.handle.mock.calls[0][0].headers.lazyUpdate).toBe(null);
```
Comment (julian-christl) [inline src/test/javascript/spec/interceptor/auth.interceptor.spec.ts:66]: Why is this line interesting to us? I have never seen it used before.

File: src/test/javascript/spec/interceptor/errorhandler.interceptor.spec.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        jest.restoreAllMocks();
+    });
+
+    it('should broadcast http errors', () => {
+        const error = new HttpErrorResponse({ status: 400 });
+        const mockHandler = {
+            handle: () => throwError(() => error),
+        };
+
+        errorHandlerInterceptor.intercept(null as any, mockHandler).subscribe();
```
Comment (julian-christl) [inline src/test/javascript/spec/interceptor/errorhandler.interceptor.spec.ts:29]: use a real value + other occurrences

File: src/test/javascript/spec/interceptor/errorhandler.interceptor.spec.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+        expect(eventManagerMock.broadcast).toHaveBeenCalledWith({
+            name: 'artemisApp.httpError',
+            content: error,
+        });
+    });
+
+    it('should not broadcast an http error if status is 401 and message is empty', () => {
```
Comment (julian-christl) [inline src/test/javascript/spec/interceptor/errorhandler.interceptor.spec.ts:37]: I think it makes sense to add a positive 401 check as well. Right now these tests would also succeed if a 401 never gets broadcasted

File: src/test/javascript/spec/interceptor/errorhandler.interceptor.spec.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    it('should broadcast http errors', () => {
+        const error = new HttpErrorResponse({ status: 400 });
+        const mockHandler = {
+            handle: () => throwError(() => error),
+        };
+
+        errorHandlerInterceptor.intercept(null as any, mockHandler).subscribe();
+
+        expect(eventManagerMock.broadcast).toHaveBeenCalledWith({
```
Comment (julian-christl) [inline src/test/javascript/spec/interceptor/errorhandler.interceptor.spec.ts:31]:
```suggestion
        expect(eventManagerMock.broadcast).toHaveBeenCalledTimes(1);
        expect(eventManagerMock.broadcast).toHaveBeenCalledWith({
```

File: src/test/javascript/spec/interceptor/notification.interceptor.spec.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                            'X-app-alert': 'Test',
+                            'X-app-params': 'ABC',
+                        }),
+                    }),
+                ),
+        } as any as HttpHandler;
+
+        notificationInterceptor.intercept(requestMock, mockHandler).subscribe();
+
+        expect(alertServiceMock.success).toHaveBeenCalledWith('Test', { param: 'ABC' });
```
Comment (julian-christl) [inline src/test/javascript/spec/interceptor/notification.interceptor.spec.ts:39]:
```suggestion
        expect(alertServiceMock.success).toHaveBeenCalledTimes(1);
        expect(alertServiceMock.success).toHaveBeenCalledWith('Test', { param: 'ABC' });
```

**Review by julian-christl — CHANGES_REQUESTED** [review]

File: src/test/javascript/spec/interceptor/auth.interceptor.spec.ts (hunk 3 above)
Comment (coaxial1) [reply src/test/javascript/spec/interceptor/auth.interceptor.spec.ts:66]: It's a way to check what the clone method on the request has been called with, but yeah, probably depens too much on the internal implementation of the http request. I've removed it and replaced the other lazyUpdate check with a spy on the clone call itself instead where it sets the headers 👍

File: src/test/javascript/spec/interceptor/errorhandler.interceptor.spec.ts (hunk 5 above)
Comment (coaxial1) [reply src/test/javascript/spec/interceptor/errorhandler.interceptor.spec.ts:37]: I'll let the first test check a few status codes

File: src/test/javascript/spec/interceptor/auth-expired.interceptor.spec.ts (hunk 1 above)
Comment (coaxial1) [reply src/test/javascript/spec/interceptor/auth-expired.interceptor.spec.ts]: 👍

File: src/test/javascript/spec/interceptor/errorhandler.interceptor.spec.ts (hunk 4 above)
Comment (coaxial1) [reply src/test/javascript/spec/interceptor/errorhandler.interceptor.spec.ts]: 👍

**Review by julian-christl — APPROVED** [review]
Thanks for the changes, code looks good

**Review by willergeorg — APPROVED** [review]
Code changes lgtm

**Pull request merged**